### PR TITLE
ci runner: try to reclaim disk space when usage is high

### PR DESCRIPTION
**Pull request by bduffany**
After running the action (and publishing the success/failure status for the outer invocation), attempt to save disk space by running `git gc` if disk usage exceeds 90%.

**Related issues**: N/A

### CODE REVIEW

File: enterprise/server/cmd/ci_runner/main.go [hunk 1]
```diff
@@ -2272,3 +2285,50 @@ func runCredentialHelper() error {
...
+	}
+	usage, err := diskUsageFraction()
+	if err != nil {
+		return fmt.Errorf("get disk usage: %w", err)
+	}
+	if usage < highDiskUsageThreshold {
+		return nil
+	}
+
+	// We should be in the git repo root at this point - run git gc.
```
Comment (maggie-lou) [inline enterprise/server/cmd/ci_runner/main.go:2303]: Previously we discussed running git gc at the end of every run. That should amortize the costs and make the behavior more predictable. WHat do you think of that approach instead?

Comment (bduffany) [reply enterprise/server/cmd/ci_runner/main.go:2303]: I'm not sure if the cost is actually amortized - if I run `git gc` in my local repo, it always seems to do a lot of work.

File: enterprise/server/cmd/ci_runner/main.go [hunk 2]
```diff
@@ -2272,3 +2285,50 @@ func runCredentialHelper() error {
...
+	log.Infof("Running git gc...")
+	if _, err := git(ctx, os.Stderr, "gc"); err != nil {
```
Comment (sluongng) [inline enterprise/server/cmd/ci_runner/main.go:2305]: Hmm, my concern is that with much garbage left over, git-gc could take a long time to complete. And this operation will block the PR check update, right?

I think we should default running these `git-maintenance` operations unconditionally instead:

```
git maintenance run loose-objects
git maintenance run incremental-repack
git maintenance run pack-refs
```

These commands are lighter weight and running them unconditionally each Invocation will divide out the total cleanup cost, avoiding 1 big blocking gc run (which, for larger repos, could cause OOM during packfile creation).

The caveat is that these are only available starting from Git 2.31.0

Comment (bduffany) [reply enterprise/server/cmd/ci_runner/main.go]: > this operation will block the PR check update, right?

No, this cleanup is run after the workflow BES stream is closed, and the git statuses are published by the BES server.

> I think we should default running these git-maintenance operations unconditionally instead

These commands seem to behave better when running multiple times in a row, so running unconditionally SGTM (updated the PR)

**Review by sluongng — APPROVED** [review]
LGTM

**Review by maggie-lou — APPROVED** [review]
Small comments but overall LGTM

**Review by maggie-lou — APPROVED** [review]

**Pull request merged**